man/man1/oci-network-inspector.1
.\" Process this file with
.\" groff -man -Tascii oci-network-inspector.1
.\"
.\" Copyright(c) 2018 Oracle and/or its affiliates. All rights reserved.
.\" Licensed under the Universal Permissive License v 1.0 as shown
.\" at http://oss.oracle.com/licenses/upl.
.\"
.TH OCI-NETWORK-INSPECTOR 1 "AUG 2018" Linux "User Manuals"
.SH NAME
oci-network-inspector \- display Oracle Cloud Infrastructure networking information.
.SH SYNOPSIS
.B oci-network-inspector [--help] [-C
.I COMPARTMENT_OCID
.B | --compartment
.I COMPARTMENT_OCID
.B ] [-N
.I VCN_OCID
.B | --vcn
.I VCN_OCID
.B ] [--help]
.SH DESCRIPTION
The
.B oci-network-inspector
utility is used to display networking information for a given VCN or compartment, or for the configured tenancy in the Oracle Cloud Infrastructure.  By default, all VCNs in all compartments accessible by the user will be displayed.

It lists detailed information including security list, IP addresses and associated VNICs and instances in all of the subnets of a given VCN, compartment or the tenancy.

It works on any system that has python36-oci-sdk installed and configured to authenticate with an Oracle Cloud Infrastructure API server.
.SH OPTIONS
.IP "-C COMPARTMENT_OCID, --compartment COMPARTMENT_OCID"
Specify the compartment you wish to inspect.
.IP "-N VCN_OCID, --vcn VCN_OCID"
Specify the VCN you wish to inspect.
.IP --debug
Print diagnostic messages.
.IP --help
Print a summary of the command line options.
.SH EXAMPLES
.IP "oci-network-inspector"
List all VCNs in all compartments that the user has access to, in the tenancy you configured for the Oracle Cloud Infrastructure SDK.

.IP "oci-network-inspector\ -N\ ocid1.vcn.oc1.iad.aaaaaaaaadkljfasljw3lkj4lkdlkdkj34lkjfkj34jh4kjh3knkjbkjhvoa"
List the network information for the specified VCN.

.IP "oci-network-inspector\ -C\ ocid1.tenancy.oc1..aaaaaaaasdlkjsdlkj34ljclkjcsjcslsc93,kjskdjhfskh3hjskchskjdq"
List the network information for all the VCNs in the specified compartment.

.SH DIAGNOSTICS
Return an exit status of 0 for success or 1 if an error occurred.


NEWS
Version 0.11.4
- LINUX-7035 - introduce oci-notify tool to send notification to OCI notification service

Version 0.11.3.3
- Remove python2 support
- LINUX-7986 - fix crash while printing network interfaces on BM shapes with oci-network-config command
- LINUX-7918 - fix regression on -n option of oci-kvm command
- LINUX-7918 - fix free vnic search for oci-kvm command on BM shapes
- LINUX-8011 - interpreter crashes displaying iSCSI information
- LINUX-7672 - fix for python3 byte handling issue.
- support for LVM root filesystem in oci-growfs
- multi vnic support for KVM guests
- removal of libexec/secondary_vnic_all_configure.sh, replaced by python implementation
- oci-kvm, added sanity around parameters passes as part of extra-args option
- add oci-image-migrate code
- move to by-uuid device name for libvirt storage pool build
- LINUX-8008 -  AttributeError: 'OCIVolume' object has no attribute 'HUMAN'
- LINUX-6752 - missing python OCI-SDk dependency

Version 0.10.2
- Update to use Python 3 on OL8

Version 0.10.1
- Added support of libvirt network for KVM guests
- Added support for libvir storage pools

Version 0.9.1
- Added support for Oracle Linux 8 Beta 1

Version 0.9.0
- LINUX-498 - Enhance oci-metadata  for keys to only return key value
              Added suport --value-only
- LINUX-560 - Cleanup utility not preserving permissions/ownerships (fixed)
              same as bug-29260959.
Version 0.8.0

- OLOCITOOLS-10 - added support for updating instance metadata for a specified instance.
                  $ oci-metadata [ -u|--update ] <key>=<value> ...
- OLOCITOOLS-11 - implemented method for retrieving metadata for other compute instances.
                  $ oci-metadata --instance-id <ocid>
- OLOCITOOLS-12 - implemented oci-metadata --export
                  $ oci-metadata --export -g key ..
                - enhanced metadata get by supporting path and wildcard in path.
                  $ oci-metadata -g /path/to/key -g /path/*/key
- OLOCITOOLS-28 - implemented oci-metadata --trim -g   ...
                  trim the path to the last key.
- OLOCITOOLS-24 - removed the metadata cach

- OLOCITOOLS-27 - fixed misleading error message when metadata does not include ssh_authorized_keys.

- fixed bug-28048699 - Needs to handle multiple physical NICs when creating secondary vnics
                       Added option "--nic-index" "to specify which physical NIC the vnic is assigned to.
- fixed bug-28599902 - oci-public-ip returns any public ip it found first.
                       Added an option [-a |--all] to list all public ips for this instance.
- fixed bug-28643343 - fixed most exceptions for oci config error.
- fixed bug-29026433 - oci-iscsi-config don't config secondary ip for second vnic automatically
                       when running delete ip option
- fixed bug-28840370 - local disks  should not count for "max_volumes" when creating iscsi disk
- fixed bug-29026422 - oci-iscsi-config -s " don't list boot partition for VM
- fixed bug-29118662 - oci-network-config detach vnic give wrong message

Version 0.7.1:

- fixed bug 28668447 - ocid needs to allow time for iSCSI connection to recover

Version 0.7:

- bump version for 0.7
- added oci-network-inspector to listing networking information.(OLOCITOOLS-5)
- added oci-growfs to support to grow filesystems (OLOCITOOLS-7)
  Currently only / boot volume is expendable.
- fixed bugs in oci-iscsi-config:
  1. bug-28493320: mis-list sdaa.. as partition of sda. It should be a seperate volume.
  2. bug 28615235 - different behavior for max_volumes setting for 32 or less than 32
  3. bug-28513898 - should not create a volume if it failed to attach
  4. privilege adjustment for create and destroy(root+oci), attach and detach(root)
  5. Bug 27835953 - objects can't be shown due to pagination issue.
  6. Bug 28612818 - A detached volume can't be reattached to the instance saying max limit reached
- fixed bug oci-network-config
  1. bug-28498139 - secondary ips on secondary VNICs not reachable issue.
  2. Bug 28080064 - oci-network-config -s can't list all secondary private ip
  3. Bug 28188853 - oci-network-config report error message when detach vnic
- expanded 'OCI' to 'Oracle Cloud Infrastructure' in man pages, specfiles (OLOCITOOLS-8)

Version 0.6:

- bump version for 0.6
- add support for API access through Instance Principals
- add support for root using a designated user's OCI config files and keys
- oci_utils API automatically detects authentication method to be used
- ocid can discover secondary IP addresses and CHAP user/password using
  OCI API calls, if the Python SDK is configured or Instance Principals used
- network proxy support for making SDK calls
- configuration files for ocid: /etc/oci-utils.d/*
- support configuring the various functions of ocid individually,
  including refresh frequency or turning them off completely.
- ocid saves state and restores all volumes and VNIC configuration after reboot
- oci-network-config: new option: --detach-vnic
- oci-iscsi-config: new option: --destroy-volume
- oci-utils APIs are now thread safe
- NEW tool: oci-image-cleanup - a script that runs a set of cleanup steps to
  prepare the instance for snapshot
- oci-kvm utility rejects attempts to create guests if the required
  virtualization support is not enabled in the image it is being executed on


libexec/oci-network-inspector.py
#
# Copyright (c) 2017, 2019 Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown
# at http://oss.oracle.com/licenses/upl.

"""
Script for collecting the cloud network, its subnets, and ip allocation details.
It need oci-sdk installed.
"""

import argparse
import logging
import sys

try:
    import json
except ImportError:
    import simplejson as json
from oci_utils import oci_api
# import oci_utils.oci_api
from oci_utils.exceptions import OCISDKError

__logger = logging.getLogger("oci-utils.oci-network-inspector")


def parse_args():
    """
    Parse the command line arguments and return an object representing the
    command line (as returned by argparse's parse_args()).

    Returns
    -------
        The argparse namespace.
    """
    parser = argparse.ArgumentParser(description='Utility for exploring '
                                                 'the network, its subnets '
                                                 'and ips assigned for a '
                                                 'given compartment or '
                                                 'network.', add_help=False)
    parser.add_argument('-C', '--compartment',
                        metavar='OCID',
                        action='store',
                        help='ocid of a compartment you like to explore ')
    parser.add_argument('-N', '--vcn',
                        metavar='OCID',
                        action='store',
                        help='ocid of a given Virtual Cloud Network')
    parser.add_argument('--help',
                        action='help',
                        help='Display this help')

    args = parser.parse_args()
    return args


def main():
    """
    Main

    Returns
    -------
        int
            0 on success, 1 otherwise.
    """
    args = parse_args()

    # needs the OCI SDK installed and configured
    try:
        sess = oci_api.OCISession()
        # sess = oci_utils.oci_api.OCISession()
    except OCISDKError as e:
        __logger.error("Need OCI Service to inspect the networks.\n"
                       "Make sure to install and configure "
                       "OCI Python SDK (python-oci-sdk)\n %s" % str(e))
        return 1

    comps = []
    if args.compartment:
        if args.compartment.startswith('ocid1.compartment.oc1..'):
            comp = sess.get_compartment(compartment_id=args.compartment)
            if comp is not None:
                comps.append(comp)
        else:
            comp = sess.find_compartments(args.compartment)
            if comp is not None:
                comps += comp
    else:
        comp = sess.all_compartments()
        comps += comp

    vcns = []
    if args.vcn:
        if args.vcn.startswith('ocid1.vcn.oc1.'):
            vcn = sess.get_vcn(args.vcn)
            if vcn is not None:
                vcns.append(vcn)
        else:
            vcn = sess.find_vcns(args.vcn)
            if vcn is not None:
                vcns += vcn
    else:
        # get all vcns for the compartment.
        for comp in comps:
            comp_vcns = comp.all_vcns()
            for vcn in comp_vcns:
                vcn.set_compartment_name(comp.get_display_name())
            vcns += comp_vcns

    if len(vcns) == 0:
        __logger.error("VCN not found: %s\n" % args.vcn)
        return 1

    comp_ocid = None
    for vcn in vcns:
        _compartment = vcn.get_compartment()
        if _compartment.get_ocid() != comp_ocid:
            print ""
            print "Compartment: %s (%s)" % \
                  (_compartment.get_display_name(), _compartment.get_ocid())
            comp_ocid = _compartment.get_ocid()
        print ""
        print "  vcn: %s " % vcn.get_display_name()
        sll = vcn.all_security_lists()
        for _, value in sll.items():
            value.print_security_list("    ")

        for subnet in vcn.all_subnets():
            print ""
            print "     Subnet: %s Availibility domain: %s" % (
                subnet.get_display_name(), subnet.get_availability_domain())
            print "         Cidr_block: %s Domain name: %s" % \
                  (subnet.get_cidr_block(), subnet.get_domain_name())

            for sl_id in subnet.get_security_list_ids():
                try:
                    sll.get(sl_id).print_security_list("       ")
                except Exception as e:
                    __logger.error("The security list %s is not in the VCN's "
                                   "list. \nException:%s" % (sl_id, e))

            '''
            vnics = subnet.all_vnics()
            for vnic in vnics:
                primary = "sec"
                if vnic.is_primary():
                    primary = "primary"
                print "      Vnic: %s(%s)(%s)" % (vnic.data.display_name,
                    primary, vnic.get_ocid())
                if subnet.data.prohibit_public_ip_on_vnic == False:
                    print "        Public IP: %s" % vnic.get_public_ip()
                print "        Private IP: %s (primary)" % vnic.get_private_ip()
                for ip in vnic.all_private_ips():
                    print "        Private IP: %s" % ip
            '''
            for ip in subnet.all_private_ips_with_primary():
                primary = ""
                if ip.is_primary():
                    primary = "primary"
                print "       Private IP: %s(%s) Host: %s" % \
                      (ip.get_address(), primary, ip.get_hostname())
                vnic = ip.get_vnic()
                if vnic:
                    print "         Vnic: %s (%s)" % \
                          (vnic.get_ocid(), vnic.get_state())
                    if subnet.is_public_ip_on_vnic_allowed():
                        print "         Vnic PublicIP: %s" % \
                              vnic.get_public_ip()
                    instance = vnic.get_instance()
                    print "         Instance: %s(%s)" % \
                          (instance.get_hostname(), instance.get_state())
                    print "         Instance ocid: %s" % (instance.get_ocid())
                else:
                    vnic_id = ip.get_vnic_ocid()
                    print "         Vnic: %s(%s)" % (vnic_id, "NotFound")
                    print "         Instance: (maybe)%s(%s)" % \
                          (ip.get_display_name(), "NotFound")
    return 0


if __name__ == "__main__":
    sys.exit(main())
